bundles/org.eclipse.equinox.preferences/src/org/eclipse/core/internal/preferences/BundleStateScopeServiceFactory.java

```diff
@@ -0,0 +1,75 @@
+/*******************************************************************************
+ * Copyright (c) 2022 Christoph Läubrich and others.
+ *
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Christoph Läubrich - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.core.internal.preferences;
+
+import org.eclipse.core.internal.runtime.MetaDataKeeper;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.core.runtime.preferences.*;
+import org.eclipse.osgi.service.datalocation.Location;
+import org.osgi.framework.Bundle;
+import org.osgi.service.component.ComponentContext;
+import org.osgi.service.component.annotations.*;
+
+/**
+ * The Bundle State Scope supply bundles with two items:
+ * <ul>
+ * <li>{@link #getNode(String)} returns a node for the bundle workspace
+ * preferences</li>
+ * <li>{@link #getLocation()} returns the state location of the bundle that
+ * acquires the service</li>
+ * </ul>
+ * The service could be acquired for example with declarative services the
+ * following way:
+ *
+ * <pre>
+ * @Reference(target = "(type=bundle)")
+ * private IScopeContext scopeContext;
+ * </pre>
+ */
+@Component(service = IScopeContext.class, property = "type=bundle", scope = ServiceScope.BUNDLE)
+public class BundleStateScopeServiceFactory implements IScopeContext {
+
+	private Bundle usingBundle;
+	private IPath bundleLocation;
+
+	// this is not used, but makes sure the service is not available before a
+	// workspace url is set
+	@Reference(target = "(&(&(objectClass=org.eclipse.osgi.service.datalocation.Location)(type=osgi.instance.area))(url=*))")
+	Location workspaceLocation;
+
+	@Activate
+	void activate(ComponentContext context) {
+		usingBundle = context.getUsingBundle();
+	}
+
+	@Override
+	public String getName() {
+		return "bundle_state"; //$NON-NLS-1$
+	}
+
+	@Override
+	public IEclipsePreferences getNode(String qualifier) {
+		return (IEclipsePreferences) InstanceScope.INSTANCE.getNode(usingBundle.getSymbolicName()).node(qualifier);
+	}
+
+	@Override
+	public IPath getLocation() {
+		if (bundleLocation == null) {
+			bundleLocation = MetaDataKeeper.getMetaArea().getStateLocation(usingBundle);
```

Comment:
The way this is assigned followed by a `mkdirs` makes it possible that a thread will get the `IPath` instance without the `mkdirs` called if multiple threads call this method at the same time.  Either need a lock here around assignment to `bundleLocation`.  If you want lockless then make `bundleLocation` volatile assign to local variable `tmpLocation`, if null then create a tmpLocation first, call `mkdirs` on the tmpLocation then finally assign `tmpLocation` to `bundleLocations`.  The drawback of that is that you may call `mkdirs` multiple times, but that is OK.

Reply:
I have similar questions about just assigning the `IPath` to a final in the constructor.  Is there value in being lazy at the point we are constructing the instance?

Reply:
I think for simplify the code we can just synchronize here, this method is not expected to be called very frequent and `MetaDataKeeper.getMetaArea()` is already synchronized as well, the idea to only create it on demand is that if a bundle never access the area (e.g. just want to use the preferences) we do not create empty folders that are never used.